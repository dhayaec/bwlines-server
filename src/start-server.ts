// tslint:disable-next-line:no-import-side-effect
import 'reflect-metadata';
import { Env, redisSessionPrefix } from './constants';
// tslint:disable-next-line:no-var-requires
require('dotenv-safe').config();
import * as connectRedis from 'connect-redis';
import * as expressRateLimit from 'express-rate-limit';
import * as expressSession from 'express-session';
import { GraphQLServer } from 'graphql-yoga';
import * as helmet from 'helmet';
import * as ioredis from 'ioredis';
import * as rateLimitRedis from 'rate-limit-redis';
import { createDb } from './utils/create-db';
import { genSchema } from './utils/schema-utils';

const redisStore = connectRedis(expressSession as any);
const redis = new ioredis();

export async function startServer() {
  if (process.env.NODE_ENV === Env.test) {
    await redis.flushall();
  }

  await createDb();

  const server = new GraphQLServer({
    schema: genSchema(),
    context: ({ request }) => ({
      redis,
      url: `${request.protocol}://${request.get('host')}`,
      session: request.session,
      req: request,
    }),
  });

  server.express.use(helmet());

  server.express.use(
    new expressRateLimit({
      store: new rateLimitRedis({
        client: redis,
      }),
      windowMs: 15 * 60 * 1000, // 15 minutes
      max: 100, // limit each IP to 100 requests per windowMs
      delayMs: 0, // disable delaying - full speed until the max limit is reached
    })
  );

  server.express.use(
    expressSession({
      store: new redisStore({
        client: redis as any,
        prefix: redisSessionPrefix,
      }),
      name: 'qid',
      secret: process.env.SESSION_SECRET as string,
      resave: false,
      saveUninitialized: false,
      cookie: {
        httpOnly: true,
        secure: process.env.NODE_ENV === Env.production,
        maxAge: 1000 * 60 * 60 * 24 * 7, // 7 days
      },
    })
  );

  return server.start(
    { port: process.env.NODE_ENV === 'test' ? 4001 : 4000 },
    ({ port }) => console.log('localhost:' + port)
  );
}
