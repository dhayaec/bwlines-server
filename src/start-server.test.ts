import axios from 'axios';
import { AddressInfo, Server } from 'net';
import { startServer } from './start-server';

let server: Server;
beforeAll(async () => {
  server = await startServer();
});

afterAll(async () => {
  await server.close();
});

describe('start-server', () => {
  it('should start', async () => {
    const { port } = server.address() as AddressInfo;
    const { data } = await axios.get(`http://localhost:${port}/ping`);
    expect(data).toEqual({ message: 'pong' });
  });
});
