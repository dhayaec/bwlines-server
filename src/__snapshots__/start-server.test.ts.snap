// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start-server should start 1`] = `
Object {
  "message": "pong",
}
`;
